addons/godot_rl_agents/sync.gd

```diff
@@ -2,165 +2,223 @@ extends Node
-		
-			var reply = {
-				"type": "reset",
-				"obs": obs
-			}
+			var obs = _get_obs_from_agents(agents_training)
+
+			var reply = {"type": "reset", "obs": obs}
 			_send_dict_as_json_message(reply)
 			# this should go straight to getting the action and setting it checked the agent, no need to perform one phyics tick
-			get_tree().set_pause(false) 
+			get_tree().set_pause(false)
 			return
-		
+
 		if need_to_send_obs:
 			need_to_send_obs = false
 			var reward = _get_reward_from_agents()
 			var done = _get_done_from_agents()
 			#_reset_agents_if_done() # this ensures the new observation is from the next env instance : NEEDS REFACTOR
-			
-			var obs = _get_obs_from_agents()
-			
-			var reply = {
-				"type": "step",
-				"obs": obs,
-				"reward": reward,
-				"done": done
-			}
+
+			var obs = _get_obs_from_agents(agents_training)
+
+			var reply = {"type": "step", "obs": obs, "reward": reward, "done": done}
 			_send_dict_as_json_message(reply)
-		
+
 		var handled = handle_message()
-	
-	elif onnx_model != null:
-		var obs : Array = _get_obs_from_agents()
-	
+
+	# Inference
+	if agents_inference.size() > 0:
+		var obs: Array = _get_obs_from_agents(agents_inference)
 		var actions = []
-		for o in obs:
-			var action = onnx_model.run_inference(o["obs"], 1.0)
+
+		for agent_id in range(0, agents_inference.size()):
+			var action = agents_inference[agent_id].onnx_model.run_inference(
+				obs[agent_id]["obs"], 1.0
+			)
 			action["output"] = clamp_array(action["output"], -1.0, 1.0)
-			var action_dict = _extract_action_dict(action["output"])
+			var action_dict = _extract_action_dict(
+				action["output"], _action_space_inference[agent_id]
+			)
 			actions.append(action_dict)
-		
-		_set_agent_actions(actions) 
+
+		_set_agent_actions(actions, agents_inference)
 		need_to_send_obs = true
-		get_tree().set_pause(false) 
-		_reset_agents_if_done()	
-		
-	else:
-		_reset_agents_if_done()	
+		get_tree().set_pause(false)
 
-func _extract_action_dict(action_array: Array):
+
+func _extract_action_dict(action_array: Array, action_space: Dictionary):
 	var index = 0
 	var result = {}
-	for key in _action_space.keys():
-		var size = _action_space[key]["size"]
-		if _action_space[key]["action_type"] == "discrete":
+	for key in action_space.keys():
+		var size = action_space[key]["size"]
+		if action_space[key]["action_type"] == "discrete":
 			result[key] = round(action_array[index])
 		else:
-			result[key] = action_array.slice(index,index+size)
+			result[key] = action_array.slice(index, index + size)
 		index += size
-		
+
 	return result
 
-func _get_agents():
-	agents = get_tree().get_nodes_in_group("AGENT")
 
-func _set_heuristic(heuristic):
+func _get_agents():
+	all_agents = get_tree().get_nodes_in_group("AGENT")
+	for agent in all_agents:
+		agent = agent as AIController3D
+		var agent_inherits_mode: bool = agent.control_mode == agent.ControlModes.INHERIT_FROM_SYNC
+
+		if agent_inherits_mode:
+			match control_mode:
```

Comment:
Note: I'm considering whether agents that inherit mode should have their mode set to that of the sync node. Then this code part can be moved out into a separate method (initialize_agents or set_agent_modes).

If so, the onnx model path checking section will need a slight modification (maybe if mode == inference and model_path is empty, use the path from sync node, otherwise use the local path from AIController).